controllers/metal3.io/preprovisioningimage_controller.go

```diff
@@ -0,0 +1,237 @@
+		}
+		return ctrl.Result{}, err
+	}
+
+	changed, err := r.update(&img, log)
+
+	if k8serrors.IsNotFound(err) {
+		delay := getErrorRetryDelay(img.Status)
+		log.Info("requeuing to check for secret", "after", delay)
+		result.RequeueAfter = delay
+	}
+	if changed {
+		log.Info("updating status")
+		err = r.Status().Update(ctx, &img)
+	}
+
+	return result, err
+}
+
+func (r *PreprovisioningImageReconciler) update(img *metal3.PreprovisioningImage, log logr.Logger) (bool, error) {
+	generation := img.GetGeneration()
+
+	url, format, errorMessage := getImageURL()
+	if errorMessage != "" {
+		log.Info("no suitable image URL available", "preferredFormat", format)
+		return setError(generation, &img.Status, reasonConfigurationError, errorMessage), nil
+	}
+
+	log.Info("image URL available", "url", url, "format", format)
+	secretManager := secretutils.NewSecretManager(log, r.Client, r.APIReader)
+	secretStatus, err := getNetworkDataStatus(secretManager, img)
+	if err == nil {
+		return setImage(generation, &img.Status, url, format,
+			secretStatus, img.Spec.Architecture,
+			"Set default image"), nil
+	}
+
+	if k8serrors.IsNotFound(err) {
+		log.Info("network data Secret does not exist")
+		return setError(generation, &img.Status, reasonMissingNetworkData, "NetworkData secret not found"), err
+	}
+
+	return false, err
+}
+
+func getImageURL() (url string, format metal3.ImageFormat, errorMessage string) {
+	format = metal3.ImageFormatISO
+	if iso := os.Getenv("DEPLOY_ISO_URL"); iso != "" {
+		url = iso
+	} else {
+		errorMessage = "No DEPLOY_ISO_URL specified"
+	}
+	return
+}
+
+func getErrorRetryDelay(status metal3.PreprovisioningImageStatus) time.Duration {
+	errorCond := meta.FindStatusCondition(status.Conditions, string(metal3.ConditionImageError))
+	if errorCond == nil || errorCond.Status != metav1.ConditionTrue {
+		return 0
+	}
+
+	// exponential delay
+	delay := time.Since(errorCond.LastTransitionTime.Time) + minRetryDelay
+
+	if delay > maxRetryDelay {
+		return maxRetryDelay
+	}
+	return delay
+}
+
+func getNetworkDataStatus(secretManager secretutils.SecretManager, img *metal3.PreprovisioningImage) (metal3.SecretStatus, error) {
+	networkDataSecret := img.Spec.NetworkDataName
+	if networkDataSecret == "" {
+		return metal3.SecretStatus{}, nil
+	}
+
+	secretKey := client.ObjectKey{
+		Name:      networkDataSecret,
+		Namespace: img.ObjectMeta.Namespace,
+	}
+	secret, err := secretManager.AcquireSecret(secretKey, img, false)
+	if err != nil {
+		return metal3.SecretStatus{}, err
+	}
+
+	return metal3.SecretStatus{
+		Name:    networkDataSecret,
+		Version: secret.GetResourceVersion(),
+	}, nil
+}
+
+func setCondition(generation int64, status *metal3.PreprovisioningImageStatus,
+	cond metal3.ImageStatusConditionType, newStatus metav1.ConditionStatus,
+	time metav1.Time, reason conditionReason, message string) {
+	newCondition := metav1.Condition{
+		Type:               string(cond),
+		Status:             newStatus,
+		LastTransitionTime: time,
```

Comment:
It looks like LastTransitionTime will be determined by SetStatusCondition and doesn't need to be explicitly set - https://github.com/kubernetes/apimachinery/blob/86c0c0f8c8e2ad6601321a412cddb1f99ff6fae5/pkg/api/meta/conditions.go#L37.

Reply:
This is true. By passing a timestamp we can make sure that all of the conditions that change simultaneously get the same timestamp... which is something that probably only I care about :laughing:

Reply:
OK, Yeah if you want them all to have the same timestamp passing it in will work.